Makefile

```diff
@@ -89,47 +87,62 @@ ifneq ($(CHECKSTYLE),0)
 	PERL5LIB=lib/perlcritic:$$PERL5LIB perlcritic lib
 endif
 
-.PHONY: test
-ifeq ($(TRAVIS),true)
-test: run-tests-within-container
-else
-test: checkstyle test-with-database
-endif
-
-.PHONY: test-unit-and-integration
-test-unit-and-integration:
-	OPENQA_CONFIG= prove ${PROVE_LIB_ARGS} ${PROVE_ARGS}
+test-with-db%: 
+	@$(MAKE) $(subst test-with-db,test-with-database,$@)
 
-.PHONY: test-with-database
-test-with-database:
+test-with-db:
+	@$(MAKE) test-with-database   
+ 
+test-with-databas%:
 	test -d $(TEST_PG_PATH) && (pg_ctl -D $(TEST_PG_PATH) -s status >&/dev/null || pg_ctl -D $(TEST_PG_PATH) -s start) || ./t/test_postgresql $(TEST_PG_PATH)
-	$(MAKE) test-unit-and-integration TEST_PG="DBI:Pg:dbname=openqa_test;host=$(TEST_PG_PATH)"
+	$(MAKE) $(subst -with-database,,$@) TEST_PG="DBI:Pg:dbname=openqa_test;host=$(TEST_PG_PATH)"
 	-pg_ctl -D $(TEST_PG_PATH) stop
 
-# prepares running the tests within Docker (eg. pulls os-autoinst) and then runs the tests considering
-# the test matrix environment variables
-# note: This is supposed to run within the Docker container unlike `launch-docker-to-run-tests-within`
-#       which launches the container.
-.PHONY: run-tests-within-container
-run-tests-within-container:
-	script/run-tests-within-container
+test:
+	prove -l -v -r
+
+test%:
+	@echo PERL5OPT=$(PERL5OPT)
+	@GLOBIGNORE=$(shell paste -sd: t/unstable_tests.txt); \
+	[ $@ == test-with-d* ] || case $@ in \
+		test-t*) prove -l -v t/$(subst test-t,,$@)*.t;; \
+		test-api*) prove -l -v t/api/$(subst test-api,,$@)*.t;; \
+		test-ui*) prove -l -v t/ui/$(subst test-ui,,$@)*.t;; \
+		test-dev*) DEVELOPER_FULLSTACK=1 prove -l -v t/33-developer_mode.t;; \
+		test-sched*) SCHEDULER_FULLSTACK=1 prove -l -v t/05-scheduler-full.t;; \
+		test-full*) FULLSTACK=1 prove -l -v t/full-stack.t;; \
+		test-unstable*) for f in $(shell cat t/unstable_tests.txt); do \
+		                  prove -l -v $$f || prove -l -v $$f || prove -l -v $$f; \
```

Comment:
see above for `$PROVE_ARGS`. All in all I don't think an escaped shell switch/case within a script section of a Makefile target is nicely readable

Reply:
Agree, but I definitely prefer it to having another script file with single switch called from single place.